DUI3-DX/Connectors/Autocad/Speckle.Connectors.AutocadShared/Operations/Receive/AutocadHostObjectBuilder.cs

```diff
@@ -43,53 +53,130 @@ CancellationToken cancellationToken
+
+    // return new(bakedObjectIds, results);
+
+    var objectGraph = _traversalFunction.Traverse(rootObject).Where(obj => obj.Current is not Collection);
+
+    // POC: these are not captured by traversal, so we need to re-add them here
+    var instanceDefinitionProxies = (rootObject["instanceDefinitionProxies"] as List<object>)
+      ?.Cast<InstanceDefinitionProxy>()
+      .ToList();
+
+    var instanceComponents = new List<(string[] path, IInstanceComponent obj)>();
+    // POC: these are not captured by traversal, so we need to re-add them here
+    if (instanceDefinitionProxies != null && instanceDefinitionProxies.Count > 0)
+    {
+      var transformed = instanceDefinitionProxies.Select(proxy => (Array.Empty<string>(), proxy as IInstanceComponent));
+      instanceComponents.AddRange(transformed);
+    }
+
+    var atomicObjects = new List<(string layerName, Base obj)>();
+
+    foreach (TraversalContext tc in objectGraph)
+    {
+      var layerName = GetLayerPath(tc, baseLayerPrefix);
+      if (tc.Current is IInstanceComponent instanceComponent)
+      {
+        instanceComponents.Add((new string[] { layerName }, instanceComponent));
+      }
+      else
+      {
+        atomicObjects.Add((layerName, tc.Current));
+      }
+    }
+
+    // Stage 1: Convert atomic objects
+    Dictionary<string, List<Entity>> applicationIdMap = new();
+    var count = 0;
+    foreach (var (layerName, atomicObject) in atomicObjects)
     {
+      onOperationProgressed?.Invoke("Converting objects", (double)++count / atomicObjects.Count);
       try
       {
-        var convertedObjects = ConvertObject(tc, baseLayerPrefix, uniqueLayerNames).ToList();
+        var convertedObjects = ConvertObject(atomicObject, layerName).ToList();
+
+        if (atomicObject.applicationId != null)
+        {
+          applicationIdMap[atomicObject.applicationId] = convertedObjects;
+        }
 
         results.AddRange(
           convertedObjects.Select(
             e =>
-              new ReceiveConversionResult(Status.SUCCESS, tc.Current, e.Handle.Value.ToString(), e.GetType().ToString())
+              new ReceiveConversionResult(
+                Status.SUCCESS,
+                atomicObject,
+                e.Handle.Value.ToString(),
+                e.GetType().ToString()
+              )
           )
         );
 
         bakedObjectIds.AddRange(convertedObjects.Select(e => e.Handle.Value.ToString()));
       }
       catch (Exception ex) when (!ex.IsFatal())
       {
-        results.Add(new(Status.ERROR, tc.Current, null, null, ex));
+        results.Add(new(Status.ERROR, atomicObject, null, null, ex));
       }
     }
 
+    // Stage 2: Convert instances
+    var (createdInstanceIds, consumedObjectIds, instanceConversionResults) = _instanceObjectsManager.BakeInstances(
+      instanceComponents,
+      applicationIdMap,
+      baseLayerPrefix,
+      onOperationProgressed
+    );
+
+    bakedObjectIds.RemoveAll(id => consumedObjectIds.Contains(id));
```

Comment:
You might be able to say:

```
// *maybe* more performant and a little more terse but will allocate more...
bakedObjectIds =  bakedObjectIds.Except(consumedObjectIds);
```

Also assuming this is the right way around, you're returning the list of objects you didn't consume...